src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,523 @@
+
+int
+ngx_http_lua_ffi_semaphore_post(ngx_http_lua_semaphore_t *sem,
+    int n, char **errstr)
+{
+    if (sem == NULL) {
+        *errstr = "semaphore is null";
+        return NGX_ERROR;
+    }
+
+    sem->resource_count += n;
+
+    dd("ngx_http_lua_ffi_semaphore_post semaphore: %p, resource_count: %d",
+       sem, sem->resource_count);
+
+    if (!sem->in_post_event && !ngx_queue_empty(&sem->wait_queue)) {
+
+        sem->sem_event.handler = ngx_http_lua_semaphore_handler;
+        sem->sem_event.data = sem;
+        sem->sem_event.log = sem->log;
+
+        sem->in_post_event = 1;
+        ngx_post_event((&sem->sem_event), &ngx_posted_events);
+    }
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_semaphore_wait(ngx_http_request_t *r,
+    ngx_http_lua_semaphore_t *sem, int wait_ms, u_char *errstr, size_t *errlen)
+{
+    ngx_http_lua_ctx_t           *ctx;
+    ngx_http_lua_co_ctx_t        *wait_co_ctx;
+    ngx_int_t                     rc;
+
+    if (r == NULL) {
+        *errlen = ngx_snprintf(errstr, *errlen, "request is null") - errstr;
+         return NGX_ERROR;
+    }
+
+    dd("ngx_http_lua_ffi_semaphore_wait semaphore: %p"
+       "value: %d, in_post_event: %d",
+       sem, sem->resource_count, sem->in_post_event);
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        *errlen = ngx_snprintf(errstr, *errlen, "ctx is null") - errstr;
+        return NGX_ERROR;
+    }
+
+    rc = ngx_http_lua_ffi_check_context(ctx, (NGX_HTTP_LUA_CONTEXT_REWRITE
+                                        | NGX_HTTP_LUA_CONTEXT_ACCESS
+                                        | NGX_HTTP_LUA_CONTEXT_CONTENT
+                                        | NGX_HTTP_LUA_CONTEXT_TIMER),
+                                        errstr, errlen);
+
+    if (rc == NGX_DECLINED) {
+        return NGX_ERROR;
+    }
+
+    if (!sem->in_post_event && sem->resource_count > 0) {
+         sem->resource_count--;
+         return NGX_OK;
+    }
+
+    if (wait_ms == 0) {
+        return NGX_DECLINED;
+    }
+
+    sem->wait_count++;
+    wait_co_ctx = ctx->cur_co_ctx;
+
+    wait_co_ctx->sleep.handler = ngx_http_lua_semaphore_timeout_handler;
+    wait_co_ctx->sleep.data = ctx->cur_co_ctx;
+    wait_co_ctx->sleep.log = r->connection->log;
+
+    ngx_add_timer(&wait_co_ctx->sleep, (ngx_msec_t) wait_ms);
+
+    dd("ngx_http_lua_ffi_semaphore_wait add timer coctx:%p wait: %d(ms)",
+        wait_co_ctx, wait_ms);
+
+    ngx_queue_insert_tail(&sem->wait_queue, &wait_co_ctx->sem_wait_queue);
+    wait_co_ctx->data = sem;
+    wait_co_ctx->cleanup = ngx_http_lua_semaphore_cleanup;
+
+    return NGX_AGAIN;
+}
+
+
+static void
+ngx_http_lua_semaphore_cleanup(void *data)
+{
+    ngx_http_lua_co_ctx_t          *coctx = data;
+    ngx_queue_t                    *q;
+    ngx_http_lua_semaphore_t       *sem;
+
+    sem = coctx->data;
+
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, sem->log, 0,
+                   "ngx_http_lua_semaphore_cleanup");
```

Comment:
Usually we do not use raw C function names in the debug log message. Traditionally, we use something like "http lua semaphore cleanup" as the message body here.